apisix/plugins/hmac-auth.lua

```diff
@@ -0,0 +1,303 @@
+
+    return consumer
+end
+
+
+local function generate_signature(ctx, secret_key, params)
+    local canonical_uri = ctx.var.uri
+    local canonical_query_string = ""
+    local request_method = ngx_req.get_method()
+    local args = ngx_req.get_uri_args()
+
+    if canonical_uri == "" then
+        canonical_uri = "/"
+    end
+
+    if type(args) == "table" then
+        local keys = {}
+        local query_tab = {}
+
+        for k, v in pairs(args) do
+            core.table.insert(keys, k)
+        end
+        core.table.sort(keys)
+
+        for _, key in pairs(keys) do
+            local param = args[key]
+            if type(param) == "table" then
+                for _, vval in pairs(param) do
+                    core.table.insert(query_tab, escape_uri(key) .. "=" .. escape_uri(vval))
+                end
+            else
+                core.table.insert(query_tab, escape_uri(key) .. "=" .. escape_uri(param))
+            end
+        end
+        canonical_query_string = core.table.concat(query_tab, "&")
+    end
+
+    local req_body = core.request.get_body()
+    req_body = req_body or ""
+
+    local signing_string = request_method .. canonical_uri
+                            .. canonical_query_string .. req_body
+                            .. params.access_key .. params.timestamp
+                            .. secret_key
+
+    return hmac_funcs[params.algorithm](secret_key, signing_string)
+end
+
+
+local function validate(ctx, params)
+    if not params.access_key or not params.signature then
+        return nil, {message = "access key or signature missing"}
+    end
+
+    local consumer, err = get_consumer(params.access_key)
+    if err then
+        return nil, err
+    end
+
+    local conf = consumer.auth_conf
+    if conf.algorithm ~= params.algorithm then
+        return nil, {message = "algorithm " .. params.algorithm .. " not supported"}
+    end
+
+    core.log.info("conf.clock_skew: ", conf.clock_skew)
+    if conf.clock_skew and conf.clock_skew > 0 then
+        local diff = abs(ngx_time() - params.timestamp)
+        core.log.info("conf.diff: ", diff)
+        if diff > conf.clock_skew then
+          return nil, {message = "Invalid timestamp"}
+        end
+    end
+
+    local secret_key          = conf and conf.secret_key
+    local request_signature   = ngx_decode_base64(params.signature)
+    local generated_signature = generate_signature(ctx, secret_key, params)
+
+    core.log.info("request_signature: ", request_signature,
+        " generated_signature: ", generated_signature)
+
+    if request_signature ~= generated_signature then
+        return nil, {message = "Invalid signature"}
+    end
+
+    return consumer
+end
+
+local function get_params(ctx)
+    local params = {}
+    local local_conf = core.config.local_conf()
+    local signature_key = SIGNATURE_KEY
+    local algorithm_key = ALGORITHM_KEY
+    local timestamp_key = TIMESTAMP_KEY
+    local access_key = ACCESS_KEY
```

Comment:
Here again define local variables to receive the constants defined above, is it for better coding style?

Reply:
I think the current code style is fine. Do you have any suggestions? @Firstsawyou